OsmAnd/src/net/osmand/plus/mapmarkers/adapters/MapMarkersGroupsAdapter.java

```diff
@@ -7,94 +7,188 @@
 
 import net.osmand.data.LatLon;
 import net.osmand.plus.IconsCache;
+import net.osmand.plus.MapMarkersHelper;
 import net.osmand.plus.MapMarkersHelper.MapMarker;
 import net.osmand.plus.OsmandApplication;
+import net.osmand.plus.OsmandSettings;
 import net.osmand.plus.R;
 import net.osmand.plus.activities.MapActivity;
+import net.osmand.plus.dashboard.DashLocationFragment;
 
 import java.text.SimpleDateFormat;
 import java.util.ArrayList;
 import java.util.Calendar;
+import java.util.Collections;
+import java.util.Comparator;
 import java.util.Date;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Locale;
+import java.util.Map;
 
 public class MapMarkersGroupsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
 
 	private static final int HEADER_TYPE = 1;
 	private static final int MARKER_TYPE = 2;
+	private static final int SHOW_HIDE_HISTORY_TYPE = 3;
 
 	private static final int TODAY_HEADER = 56;
 	private static final int YESTERDAY_HEADER = 57;
 	private static final int LAST_SEVEN_DAYS_HEADER = 58;
 	private static final int THIS_YEAR_HEADER = 59;
 
+	private MapActivity mapActivity;
 	private OsmandApplication app;
     private List<Object> items = new ArrayList<>();
+	private List<MapMarkersGroup> groups;
     private boolean night;
     private int screenOrientation;
     private LatLon location;
     private Float heading;
     private boolean useCenter;
 
     public MapMarkersGroupsAdapter(MapActivity mapActivity) {
-		this.app = mapActivity.getMyApplication();
+		this.mapActivity = mapActivity;
+		app = mapActivity.getMyApplication();
         night = !mapActivity.getMyApplication().getSettings().isLightContent();
 		createDisplayGroups();
     }
 
     public void createDisplayGroups() {
 		items.clear();
```

Comment:
what reason to clear items?

Reply:
createDisplayGroups() was called when user opens groups tab. So it was necessary to clear list before adding new items to it. Now this method is called only once when initializing adapter.